pkg/util/maps/maps.go

```diff
@@ -35,24 +35,48 @@ func Clone[K comparable, V any, S ~map[K]V](s S) S {
 	return ret
 }
 
-// Merge merges a and b while resolving the conflicts by calling resolveConflict
-func Merge[K comparable, V any, S ~map[K]V](a, b S, resolveConflict func(a, b V) V) S {
+func join[K comparable, V any, M ~map[K]V](a, b M, commonKeyVal func(a, b V) V, exclusive bool) M {
 	if a == nil {
+		if exclusive {
+			return nil
+		}
 		return Clone(b)
 	}
 
-	ret := Clone(a)
+	if b == nil {
+		if exclusive {
+			return nil
+		}
+		return Clone(a)
+	}
+
+	var ret M
+	if exclusive {
+		ret = make(M, len(a))
+	} else {
+		ret = Clone(a)
+	}
 
 	for k, v := range b {
-		if _, found := ret[k]; found {
-			ret[k] = resolveConflict(ret[k], v)
-		} else {
+		if _, found := a[k]; found {
+			ret[k] = commonKeyVal(a[k], v)
+		} else if !exclusive {
 			ret[k] = v
 		}
 	}
 	return ret
 }
 
+// Merge merges a and b while resolving the conflicts by calling resolveConflict
+func Merge[K comparable, V any, S ~map[K]V](a, b S, resolveConflict func(a, b V) V) S {
```

Comment:
```suggestion
func Merge[K comparable, V any, S ~map[K]V](a, b S, val func(a, b V) V) S {
```
For consistency with the `Intersect`.

Reply:
`resolveConflict` makes better sense in the context of merge, let's keep it as is.

Reply:
Should we call it then `resolveConflict` in both places? It seems to play exactly the same role in both functions

Reply:
In case of `Intersect` you always expect values coming from both sides.

Reply:
Values coming from both sides is specifically what I would call conflict resolution (if we want to go this way).

By analogy to diffs, if the line changes are coming from only one side there is no conflict, just accept the change. If they are coming from two sides, then we need to solve the conflict.